### ISSUE TITLE
DID parallelization of loop domain

### ISSUE THREAD

**Issue by cowanmeg** [NVIDIA/Fuser#2563]
Currently, DID parallelization is only supported on the logical domain which requires us to manually add extra axes in the logical domain when parallelizing Fusions across multiple devices. 

This has caused some PRs to introduce temporary hacks to work around logical DID axes or code that needs to be modified which is tracked here.

**Comment by wujingyue** [comment]
As we discussed in the meeting, here's a breakdown of the task: 
- [x] https://github.com/NVIDIA/Fuser/issues/3282. ETC: 3 weeks. Assigned to @wujingyue 
- [x] Fix communication lowering to support DID split. ETC: 2 weeks. Assigned to @wujingyue 
- [x] Fix ATen-evaluated nodes, e.g. linear, matmul and sdpa. ETC: 2 weeks
- [ ] Fix all schedulers. According to [the nsys traces](https://docs.google.com/spreadsheets/d/1XvKcvhrIqmxsoQMErH8-pcWVBOwlmbHNJlfQBS4NeVM/edit?gid=0#gid=0) @cowanmeg collected, we'll need to fix at least pointwise, *_persistent (aka normalization), transpose and reduction. FWIW, forward needs pointwise and persistent only. ETC: 1 month
- [x] Fix sharding propagation to propagate DID splits, e.g., using TransformPropagator. ETC: 1 month 
- [x] Fix other sharding-related pre-segmenter passes, e.g., insert_reshardings and reorder_sharded_axis. One example is to decompose a DID-involving reduction (e.g. Reduction, Matmul and Linear) into multiple ops. This is conceptually similar to the existing reduction rfactoring. ETC:  1 month
- [ ] Remove tests that rely on logical split of DID and clean up dead code. ETC: 2 weeks

These subtasks are roughly execution ordered. #3282 pretty much blocks everything else, so it's currently my top priority. After that's done, some other tasks (e.g. 2-4) can be parallelized as long as we limit the scope of testing. 

cc @naoyam, @cowanmeg and @samnordmann to take a look

**Comment by cowanmeg** [comment]
Yes, I actually experimented a bit with that in this very old and out of date PR, https://github.com/NVIDIA/Fuser/pull/2181
When we explicitly set the allocation domain of the sharded tensors to the loop domain (before the schedulers applied intra-device parallelization) allocation worked.

**Comment by wujingyue** [comment]
> It's probably necessary to update the allocation logic of gmem tensors.

I believe it has been fixed: https://github.com/NVIDIA/Fuser/blob/85c22a2af3fc756d9b301c93ac29b2bba8abf2af/csrc/runtime/allocations.cpp#L737-L746. Anyhow, even if it hasn't, we will be enforced to fix it along the way of fixing the subtasks on the list.

**Comment by samnordmann** [comment]
Thanks for creating this issue so we can keep track!

> the logical domain is used to determine the allocation size, which would need to be changed to the loop domain.
This is actually just one example of loop vs logical. I could imagine there would be some other cases as well.

@naoyam agreed! 

I also think there are a bunch of conventions we also need to discuss regarding the fact that, if using split/merge on sharded axis, we will need some operations/logic to operate on the loop domain and not on the logical domain.

_However: an alternative would be to not use TensorView::split but a ViewOp proper Fusion IR operation to split/merge the axis. This way we could avoid operating on the loop domain. Is it correct? Wdyt? That might be a better direction._

To try to illustrate more explicitly why, if using splits, the allocation will need to look at the loop domain, let me consider the simplest example possible, without any resharding:

```cpp
DeviceMesh mesh = {0,1};

auto fusion = std::make_unique<Fusion>();
FusionGuard fg(fusion.get());

TensorView* tv0 = makeConcreteTensor(128);
TensorView* tv1 = set(tv0);
fusion->addInput(tv0);
fusion->addOutput(tv1);

tv0->setDeviceMesh(mesh);
tv1->setDeviceMesh(mesh);

tv0->split(0, num_devices, false);
tv1->split(0, num_devices, false);
```

the domains are as follows:
- root/logical [128]
- loop [2, 64]
- allocation [1, 64]

The fusion should expect a tensor of shape [1,64] consistent with the allocation domain, which can't be deduced from the logical domain alone

**Comment by wujingyue** [comment]
Reassigning this to @Priya2698 as she's actively working on this

**Comment by Priya2698** [comment]
I am closing this since the building blocks are complete.
Let's open targeted issues for other work such as context parallel etc when needed.
